proposals/nnnn-safe-shared-contiguous-storage.md

```diff
@@ -0,0 +1,1197 @@
+
+A `Span` provided by container represents a borrow of that container. `Span` can therefore provide simultaneous access to a non-copyable container, and can help avoid unwanted copies of copyable containers. Note that `Span` is not a replacement for a copyable container with owned storage; see [future directions](#Directions) for more details ([Resizable, contiguously-stored, untyped collection in the standard library](#Bytes))
+
+`Span` is the currency type for local processing over values in contiguous memory. It is a replacement for many API currently using `Array`, `UnsafeBufferPointer`, `Foundation.Data`, etc., that do not need to escape the owning container.
+
+### `RawSpan`
+
+`RawSpan` allows sharing the contiguous internal representation for values which may be heterogeneously-typed, such as in decoders. Since it is a fully concrete type, it can achieve better performance in debug builds of client code as well as a more straightforward understanding of performance in library code.
+
+`Span<some BitwiseCopyable>` can always be converted to `RawSpan`, using a conditionally-available property.
+
+## <a name="Design"></a>Detailed design
+
+`Span<Element>` is a simple representation of a region of initialized memory.
+
+```swift
+public struct Span<Element: ~Copyable & ~Escapable>: Copyable, ~Escapable {
+  internal var _start: UnsafePointer<Element>
+  internal var _count: Int
+}
+```
+
+It provides a buffer-like interface to the elements stored in that span of memory:
+
+```swift
+extension Span where Element: ~Copyable & ~Escapable {
+  public var count: Int { get }
+  public var isEmpty: Bool { get }
+
+  subscript(_ position: Int) -> Element { _read }
+}
+```
+
+Note that `Span` does _not_ conform to `Collection`. This is because `Collection`, as originally conceived and enshrined in existing source code, assumes pervasive copyability and escapability of the `Collection` itself as well as of element type. In particular a subsequence of a `Collection` is semantically a separate value from the instance it was derived from. In the case of `Span`, a sub-span representing a subrange of its elements _must_ have the same lifetime as the `Span` from which it originates. Another proposal will consider collection-like protocols to accommodate different combinations of `~Copyable` and `~Escapable` for the collection and its elements.
+
+`Span`s representing subsets of consecutive elements can be extracted out of a larger `Span` with an API similar to the `extracting()` functions recently added to `UnsafeBufferPointer` in support of non-copyable elements:
+
+```swift
+extension Span where Element: ~Copyable & ~Escapable {
+  public func extracting(_ bounds: Range<Int>) -> Self
```

Comment:
I think it will work against us to include any APIs whose semantics require talking about lifetime dependencies. It will open of a line of discussion that we can't resolve and will cause delays. If we're going to split Span into two pitches, then let's restrict the lifetime dependence discussion to the second pitch and focus on that second pitch as soon as possible.

Reply:
These are moved to a future directions item.